File: apps/api/src/domain/dispatch.ts
import { ValidationError } from '../common/errors';
import { isAssignable, type DriverStatus } from './driver-status';
import type { RideStatus } from './ride-status';

/**
 * Which driver may take which ride, and which ride dispatch should look at
 * next. Pure: no database, no clock beyond the `now` it is handed.
 *
 * The MVP assigns manually — a dispatcher picks from a queue (R1). This file is
 * the part that must not be a matter of dispatcher judgement: the wheelchair
 * rule in particular, because the failure mode is a patient in a wheelchair
 * meeting a saloon car at the kerb, twenty minutes before a cardiology
 * appointment, with no second vehicle available. Auto-assignment, when it
 * lands, will rank candidates this function has already filtered.
 */

export interface DriverCandidate {
  driverId: string;
  displayName: string;
  status: DriverStatus;
  onShift: boolean;
  vehicleIsWheelchairAccessible: boolean;
  /** Rides already on this driver that have not reached a terminal state. */
  activeRideCount: number;
}

export interface RideDemand {
  rideId: string;
  status: RideStatus;
  scheduledPickupAt: Date;
  wheelchairRequired: boolean;
}

/** Statuses a dispatcher is being asked to act on. */
const AWAITING: readonly RideStatus[] = [
  'requested',
  'awaitingAssignment',
  'reassignmentRequired',
];

export function isAwaitingDispatch(status: RideStatus): boolean {
  return AWAITING.includes(status);
}

export type IneligibilityReason =
  | 'notApproved'
  | 'offShift'
  | 'noAccessibleVehicle'
  | 'alreadyOnARide';

export interface Eligibility {
  eligible: boolean;
  /** Every reason, not the first — a dispatcher needs the whole picture. */
  reasons: readonly IneligibilityReason[];
}

/**
 * Reasons rather than a boolean, because a dispatcher looking at an empty
 * candidate list has to know whether the answer is "nobody is on shift" or
 * "nobody has an accessible vehicle". Those need different phone calls.
 */
export function driverEligibility(
  driver: DriverCandidate,
  ride: Pick<RideDemand, 'wheelchairRequired'>,
): Eligibility {
  const reasons: IneligibilityReason[] = [];

  if (!isAssignable(driver.status)) reasons.push('notApproved');
  if (!driver.onShift) reasons.push('offShift');
  if (ride.wheelchairRequired && !driver.vehicleIsWheelchairAccessible) {
    reasons.push('noAccessibleVehicle');
  }
  // One passenger at a time. A driver cannot be two places at once, and a
  // dispatcher under pressure will double-book if nothing stops them.
  if (driver.activeRideCount > 0) reasons.push('alreadyOnARide');

  return { eligible: reasons.length === 0, reasons };
}

export function eligibleDrivers(
  drivers: readonly DriverCandidate[],
  ride: Pick<RideDemand, 'wheelchairRequired'>,
): readonly DriverCandidate[] {
  return drivers.filter((driver) => driverEligibility(driver, ride).eligible);
}

/**
 * Asserts an assignment is legal. Throws rather than returning false, so a
 * caller cannot forget to check the result.
 */
export function assertAssignable(
  driver: DriverCandidate,
  ride: Pick<RideDemand, 'wheelchairRequired'>,
): void {
  const { eligible, reasons } = driverEligibility(driver, ride);
  if (eligible) return;

  throw new ValidationError(
    EXPLANATIONS[reasons[0] as IneligibilityReason],
    'driverId',
  );
}

const EXPLANATIONS: Record<IneligibilityReason, string> = {
  notApproved: 'That driver has not been approved to carry passengers.',
  offShift: 'That driver is not on shift.',
  noAccessibleVehicle: 'This trip needs a wheelchair-accessible vehicle.',
  alreadyOnARide: 'That driver is already on a trip.',
};

export type DispatchUrgency = 'overdue' | 'imminent' | 'soon' | 'later';

/**
 * How close a ride is to needing a car, measured against its pickup time.
 *
 * The bands are what a queue is sorted and coloured by. `overdue` is its own
 * band rather than the top of `imminent` because a ride whose pickup time has
 * passed with nobody assigned is a *failure already in progress*, not an
 * urgent task — somebody is standing in a hallway waiting.
 */
export function dispatchUrgency(pickupAt: Date, now: Date): DispatchUrgency {
  const minutes = (pickupAt.getTime() - now.getTime()) / 60_000;

  if (minutes < 0) return 'overdue';
  if (minutes <= 30) return 'imminent';
  if (minutes <= 120) return 'soon';
  return 'later';
}

const URGENCY_ORDER: Record<DispatchUrgency, number> = {
  overdue: 0,
  imminent: 1,
  soon: 2,
  later: 3,
};

/**
 * The dispatch queue, ordered.
 *
 * Overdue first, then by pickup time. Deliberately *not* by when the request
 * arrived: a ride booked this morning for 4pm is not more urgent than one
 * booked five minutes ago for 2pm, and a first-in-first-out queue quietly
 * optimises for the dispatcher's sense of fairness rather than for the person
 * waiting.
 */
export function dispatchQueue<T extends RideDemand>(
  rides: readonly T[],
  now: Date,
): readonly (T & { urgency: DispatchUrgency })[] {
  return rides
    .filter((ride) => isAwaitingDispatch(ride.status))
    .map((ride) => ({ ...ride, urgency: dispatchUrgency(ride.scheduledPickupAt, now) }))
    .sort((a, b) => {
      const byUrgency = URGENCY_ORDER[a.urgency] - URGENCY_ORDER[b.urgency];
      if (byUrgency !== 0) return byUrgency;
      return a.scheduledPickupAt.getTime() - b.scheduledPickupAt.getTime();
    });
}

File: apps/ops_console/lib/domain/dispatch.dart
/// The dispatch rules, as the console renders them. Mirrors
/// apps/api/src/domain/dispatch.ts and apps/api/src/domain/driver-status.ts.
///
/// Like `lib/domain/` in the family app, this copy **decides nothing**. The
/// server asserts every assignment — `assertAssignable` throws rather than
/// returning a boolean a caller can forget to read — and this file exists so a
/// dispatcher sees *why* a driver is unavailable before they tap, rather than
/// after. Deleting it would change no outcome except which rows look tappable.
///
/// The distinction matters more here than on the family side. A dispatcher
/// under pressure at 8am is not the last line of defence against a patient in
/// a wheelchair meeting a saloon car at the kerb; the server is. This is the
/// part that makes the queue legible enough that they rarely find out.
library;

/// How close a ride is to needing a car.
enum DispatchUrgency {
  /// The pickup time has passed with nobody assigned.
  overdue,
  imminent,
  soon,
  later;

  static DispatchUrgency? tryParse(String? value) {
    for (final urgency in DispatchUrgency.values) {
      if (urgency.name == value) return urgency;
    }
    return null;
  }

  String get wire => name;

  /// What the band is called on screen.
  ///
  /// `overdue` is deliberately not "very urgent". It is a failure already in
  /// progress — somebody is standing in a hallway waiting — and a dispatcher
  /// scanning a queue needs that to read differently from "do this next".
  String get label => switch (this) {
    DispatchUrgency.overdue => 'Overdue',
    DispatchUrgency.imminent => 'Within 30 min',
    DispatchUrgency.soon => 'Within 2 hours',
    DispatchUrgency.later => 'Later',
  };

  /// Sort weight, matching `URGENCY_ORDER` on the server.
  int get order => switch (this) {
    DispatchUrgency.overdue => 0,
    DispatchUrgency.imminent => 1,
    DispatchUrgency.soon => 2,
    DispatchUrgency.later => 3,
  };

  /// Whether the row should be shown as demanding attention rather than
  /// merely scheduled.
  bool get isPressing =>
      this == DispatchUrgency.overdue || this == DispatchUrgency.imminent;
}

/// Why a driver cannot take a particular trip.
///
/// Every reason is carried, never just the first, because a dispatcher looking
/// at an empty candidate list has to know whether the answer is "nobody is on
/// shift" or "nobody has an accessible vehicle". Those need different phone
/// calls, and collapsing them to "no drivers available" turns a two-minute fix
/// into a cancelled appointment.
enum IneligibilityReason {
  notApproved,
  offShift,
  noAccessibleVehicle,
  alreadyOnARide;

  static IneligibilityReason? tryParse(String? value) {
    for (final reason in IneligibilityReason.values) {
      if (reason.name == value) return reason;
    }
    return null;
  }

  String get wire => name;

  String get label => switch (this) {
    IneligibilityReason.notApproved => 'Not approved to carry passengers',
    IneligibilityReason.offShift => 'Not on shift',
    IneligibilityReason.noAccessibleVehicle =>
      'No wheelchair-accessible vehicle',
    IneligibilityReason.alreadyOnARide => 'Already on a trip',
  };

  /// What a dispatcher could actually do about it, where there is something.
  ///
  /// `noAccessibleVehicle` has no suggestion on purpose: the answer is another
  /// vehicle, not another decision, and offering "assign anyway" is exactly
  /// the affordance the server refuses.
  String? get remedy => switch (this) {
    IneligibilityReason.offShift => 'Put them on shift from the roster',
    IneligibilityReason.notApproved => 'Approve them from the roster',
    IneligibilityReason.alreadyOnARide =>
      'Wait for their current trip to finish',
    IneligibilityReason.noAccessibleVehicle => null,
  };
}

/// The lifecycle of a driver at an operator.
enum DriverStatus {
  invited,
  pendingApproval,

  /// May be assigned rides. The only status that occupies a billable seat.
  approved,
  suspended,
  offboarded;

  static DriverStatus? tryParse(String? value) {
    for (final status in DriverStatus.values) {
      if (status.name == value) return status;
    }
    return null;
  }

  String get wire => name;

  String get label => switch (this) {
    DriverStatus.invited => 'Invited',
    DriverStatus.pendingApproval => 'Awaiting approval',
    DriverStatus.approved => 'Approved',
    DriverStatus.suspended => 'Suspended',
    DriverStatus.offboarded => 'Offboarded',
  };

  /// Whether this driver counts towards the per-driver subscription.
  ///
  /// One definition on each side, and the server's is the one that bills. Shown
  /// here so an admin about to approve somebody can see it costs money before
  /// they tap, rather than discovering it on an invoice.
  bool get occupiesSeat => this == DriverStatus.approved;

  bool get isAssignable => this == DriverStatus.approved;

  bool get isTerminal => this == DriverStatus.offboarded;
}

/// The transitions the server's state machine allows.
///
/// Mirrored so a menu offers only the moves that will succeed. Note the
/// absence of `pendingApproval → …` straight to nothing else and, more
/// importantly, that there is **no edge from `invited` to `approved`**:
/// approval is a decision about documents somebody submitted, and an operator
/// that can approve an empty file has an onboarding control that does nothing.
const Map<DriverStatus, List<DriverStatus>> allowedDriverTransitions = {
  DriverStatus.invited: [DriverStatus.pendingApproval, DriverStatus.offboarded],
  DriverStatus.pendingApproval: [
    DriverStatus.approved,
    DriverStatus.suspended,
    DriverStatus.offboarded,
  ],
  DriverStatus.approved: [DriverStatus.suspended, DriverStatus.offboarded],
  DriverStatus.suspended: [DriverStatus.approved, DriverStatus.offboarded],
  DriverStatus.offboarded: [],
};

List<DriverStatus> nextStatusesFor(DriverStatus from) =>
    allowedDriverTransitions[from] ?? const [];

bool canTransitionDriver(DriverStatus from, DriverStatus to) =>
    nextStatusesFor(from).contains(to);

/// Whether moving to [to] requires the dispatcher to say why.
///
/// Suspension and offboarding both end somebody's ability to earn, and a
/// record of one with no reason attached is unanswerable three months later
/// when they ask. The server accepts a reason on any transition; the console
/// insists on one for these two.
bool transitionNeedsReason(DriverStatus to) =>
    to == DriverStatus.suspended || to == DriverStatus.offboarded;
